plugins/state_monitor.py

```diff
@@ -0,0 +1,238 @@
+"""
+Plugin to monitor state variables from the charge controller.
+"""
+
+# Standard library imports
+from pathlib import Path
+import urllib.request  # for SlackSender
+import urllib.parse  # for SlackSender
+from configparser import ConfigParser  # for SlackSender
+
+# Local imports
+import brokkr.pipeline.base
+import brokkr.pipeline.decode
+
+
+class StateMonitor(brokkr.pipeline.base.OutputStep):
+    """ Handle notifications for changes in state variables. """
+
+    def __init__(self, method=None, power_delim=1,
+                 slack_channel=None,
+                 slack_key_file=None,
+                 **output_step_kwargs):
+        """
+
+        Parameters
+        ----------
+        method : str
+            The method of how we're going to send notifications. If `None`, then
+            we'll only log them.
+        power_delim : numeric
+            The delimiter between normal power and low power. If power falls below this
+            value, it is considered low and a notification will be generated.
+        slack_channel : str
+            If `method=='slack'`, then this is the channel in Slack to post notifications.
+        slack_key_file : str
+            If `method=='slack'`, then the path to the file that contains the Slack key
+        output_step_kwargs : **kwargs, optional
+            Keyword arguments to pass to the OutputStep constructor.
+
+        Returns
+        -------
+        None.
+        """
+        # Pass arguments to superclass init
+        super().__init__(**output_step_kwargs)
+
+        # Setup simpleeval parser and class initial state
+        self._previous_data = None
+        self.power_delim = power_delim
+        self.method_cls = None  # Make sure we "initialize" the attribute
+        if method == 'slack':
+            try:
+                self.method_cls = SlackSender(slack_key_file, slack_channel,
+                                              logger=self.logger)
+                # todo: pass in logger instance or use logging.getLogger(__name__)?
+            except FileNotFoundError:
+                self.logger.error("File not found error. Is the Slack key file in right place?")
+            except Exception as exp:  # if anything goes wrong, don't set the sender class
+                self.logger.error(f"Unknown error initializing SlackSender: {type(exp)}")
+
+    def execute(self, input_data=None):
+        """
+        Executing an action upon detection an arbitrary condition in the data.
+
+        Parameters
+        ----------
+        input_data : any, optional
```

Comment:
Your usage of it assumes its a `Mapping` of 

```suggestion
        input_data : Mapping[str, DataValue], optional
```

Since this is how you use it